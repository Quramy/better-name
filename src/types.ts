export type $DiffKey<T, U> = T extends U ? never : T;
export type $Diff<T, U> = Pick<T, $DiffKey<keyof T, keyof U>>;
export type $Optional<T> = { [P in keyof T]?: T[P] };
export type $PartialOptional<T, U> = $Diff<T, U> & $Optional<U>

export type FileId = string;

export interface FileRef {
  readonly id: FileId;
  readonly path: string;
}

export interface SourceReader {
  read(file: FileRef): Promise<string>;
}

export interface SourceWriter {
  write(file: FileRef, source: string): Promise<void>;
}

export interface SourceRemover {
  delete(file: FileRef): Promise<void>;
}

export interface DocumentRef {
  getFile(): FileRef;
  getDoc(): DocumentEntity;
  detach(): void;
  move(newFile: FileRef): Promise<this>;
}

export interface TransformOptions {
  from: string;
  to: string;
}

export interface DocumentEntityCreateOptions {
  fileRef: FileRef,
  fileMappingOptions?: FileMappingOptions,
}

export interface DocumentEntity {
  readonly fileRef: FileRef;
  readonly isDirty: boolean;
  reader: SourceReader;
  writer: SourceWriter;
  parse(): Promise<this>;
  transformPreceding(to: string): this;
  transformFollowing(opt: TransformOptions): this;
  flush(force?: boolean): Promise<this>;
  move(newFile: FileRef): Promise<this>;
}

export type RootImportConfig = {
  rootPathSuffix?: string;
  rootPathPrefix?: string;
};

export type FileMappingOptions = {
  rootImport?: RootImportConfig[];
}

export type FindQuery = {
  start: string;
};

export interface Project {
  getProjectDir(): string;
  getFileMappingOptions(): FileMappingOptions;
  getDocumentsList(): Promise<DocumentRef[]>;
  findOne(fileId: FileId): Promise<{ found: DocumentRef | undefined, rest: DocumentRef[] }>;
  find(query: FindQuery): Promise<{ found: DocumentRef[], rest: DocumentRef[] }>;
}
