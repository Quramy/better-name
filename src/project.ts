import { parse } from "babylon";
import { File, ImportDeclaration, stringLiteral } from "babel-types";
import traverse from "babel-traverse";
import generate from "babel-generator";
import * as glob from "glob";
import { shouldBeReplaced } from "./functions";

export type $DiffKey<T, U> = T extends U ? never : T;
export type $Diff<T, U> = Pick<T, $DiffKey<keyof T, keyof U>>;
export type $Optional<T> = { [P in keyof T]?: T[P] };
export type $PartialOptional<T, U> = $Diff<T, U> & $Optional<U>

export interface Project {
  getProjectDir(): string;
  getDocumentsList(): Promise<DocumentRef[]>;
}

export interface DocumentRef {
  readonly ref?: DocumentEntity;
  detach(): void;
}

export interface DocumentEntity {
  readonly isDirty: boolean;
  parse(): Promise<this>;
  reader?: SourceReader;
  writer?: SourceWriter;
  transform(opt: TransformOptions): this;
  flush(): Promise<this>;
}

export class DefaultProject implements Project {
  private _docRefList?: DocumentRef[];

  constructor(
    private _config: AllProjectOptions,
  ) {
  }

  getProjectDir() {
    return this._config.rootDir;
  }

  async getDocumentsList() {
    if (this._docRefList) {
      return Promise.resolve(this._docRefList);
    }
    return new Promise<DocumentRef[]>((resolve, reject) => {
      glob(this._config.pattern, (err, files) => {
        if (err) return reject(err);
      });
    });
  }
}

export type AllProjectOptions = {
  rootDir: string;
  pattern: string;
};

export const defaultProjectConfig = {
  pattern: "src/**/*.js",
}

export type ProjectOptions = $PartialOptional<AllProjectOptions, typeof defaultProjectConfig>;

export function createProject(configuration: ProjectOptions) {
  const conf = { ...defaultProjectConfig, ...configuration }
  return new DefaultProject(conf);
}

export interface TransformOptions {
  from: string;
  to: string;
}

export interface SourceReader {
  read(id: string): Promise<string>;
}

export interface SourceWriter {
  write(id: string, source: string): Promise<void>;
}

export class BabylonDocmentEntity implements DocumentEntity {

  
  private _dirty: boolean = true;
  private _rawSource?: string;
  private _file?: File;

  readonly fileId: string;

  reader!: SourceReader;
  writer!: SourceWriter;

  constructor ({
    fileId,
  }: {
    fileId: string,
  }) {
    this.fileId = fileId;
  }

  get isDirty() {
    return this._dirty;
  }

  async parse() {
    if (!this._dirty && this._rawSource) {
      return this;
    } else {
      this._rawSource = await this.reader.read(this.fileId);
      this._file = parse(this._rawSource, {
        sourceType: "module",
      });
      this._dirty = false;
      return this;
    }
  }

  transform({ from, to } : TransformOptions): this {
    if (!this._file) {
      throw new Error("Call parse");
    }
    let flag = false;
    let newModuleName: string;
    traverse(this._file, {
      ImportDeclaration: (path) => {
        // if (/hogehoge/.test(path.node.source.value)) flag = true;
        const result = shouldBeReplaced({
          targetFileId: this.fileId,
          targetModuleName: path.node.source.value,
          movingFileId: from,
          toFileId: to,
        });
        if (result.hit) {
          flag = true;
          newModuleName = result.newModuleId;
        }
      },
      exit(path) {
        if (flag && path.isImportDeclaration()) flag = false;
      },
      StringLiteral: (path) => {
        if (flag && newModuleName) {
          path.replaceWith(stringLiteral(newModuleName));
          flag = false;
        }
      },
    });
    return this;
  }

  async flush() {
    if (!this._file || !this._rawSource) {
      throw new Error("Cannot flush because the source or AST is not set.");
    }
    await this.writer.write(this.fileId, generate(this._file, {}, this._rawSource).code);
    return this;
  }

}
