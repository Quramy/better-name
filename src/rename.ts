import { Project } from "./types";
import { DefaultFileRef } from "./file-util";
import * as path from "path";

function createToPath(from: string, to: string) {
  const f = path.basename(from);
  if (path.extname(to) === "") {
    return path.join(to, f);
  } else {
    return to;
  }
}

export async function rename(prj: Project, fromPath: string, toPath: string) {
  const fromPrefix = new DefaultFileRef(fromPath, prj.getProjectDir());
  const { found } = await prj.find({ start: fromPrefix.id });
  if (!found.length) return prj;
  await found.reduce(async (prev, file) => {
    await prev;
    const { found, rest } = await prj.findOne(file.getFile().id);
    const to = new DefaultFileRef(createToPath(file.getFile().path, toPath), prj.getProjectDir());
    if (found) {
      await found.getDoc().parse();
      found.getDoc().transformPreceding(to.id);
    }
    await Promise.all(rest.map(async docRef => {
      const doc = docRef.getDoc();
      await doc.parse();
      doc.transformFollowing({ from: file.getFile().id, to: to.id });
      await doc.flush();
      return;
    }));
    if (!found) return prj;
    await found.move(to);
    return prj;
  }, Promise.resolve() as Promise<any>);
  return prj;
}
