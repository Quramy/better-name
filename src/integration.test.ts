import { FileRef, SourceWriter, SourceRemover, } from "./types";
import { DefaultProject } from "./project";
import { rename } from "./rename";
import * as path from "path";

class FixtureWriter implements SourceWriter {
  contentsMap = new Map<string, string>();

  write(file: FileRef, source: string): Promise<void> {
    this.contentsMap.set(file.id, source);
    return Promise.resolve();
  }

  dump() {
    const keys = Array.from(this.contentsMap.keys()).sort();
    return keys.map(k => {
      return `
// id: ${k}

${this.contentsMap.get(k)}`;
    }).join("\n");
  }
}

class NoopRemover implements SourceRemover {
  delete() {
    return Promise.resolve();
  }
}

class TestProject extends DefaultProject {
  protected writer = new FixtureWriter();
  protected remover = new NoopRemover();

  getSnapshot() {
    return this.writer.dump();
  }
}

describe("Integration", () => {
  it("simple_babel_prj", async done => {
    const rootDir = path.join(__dirname, "../test-fixtures/simple_babel_prj");
    const prj = new TestProject({
      rootDir,
      pattern: "src/**/*.js",
    });
    try {
      await rename(prj, path.join(rootDir, "src/core/target.js"), path.join(rootDir, "src/feat/dest.js"));
      expect(prj.getSnapshot()).toMatchSnapshot();
      done();
    } catch (err) {
      done(err);
    }
  });
});
