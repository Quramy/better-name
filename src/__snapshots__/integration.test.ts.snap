// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration simple_babel_prj 1`] = `
"
// id: src/core/dependee.js

export function util() {
  console.log('this is util func');
}

// id: src/core/index.js

import { someFn } from '../feat/dest';

export default function main() {
  someFn();
}

// id: src/feat/dest.js

import { util } from '../core/dependee';

export function someFn() {
  util();
  return 'test';
}"
`;
